### DISCUSSION TITLE
Can not login with correct credentials

### DISCUSSION THREAD

**Discussion by Turab** [dani-garcia/vaultwarden#3853] [Q&A]
Version: 1.29.2
Environment: Docker deploy through Caprover

3 days after the update to 1.29.2 (and setting up push notifications), I cannot login through only some of the clients. I didn't open this as a discussion because the behavior is very much unexpected and logs the error "password incorrect" although it is correct. Because some of the clients work with that password.

Login to Opera extension with password: Works
Login to Chrome extension with password on the same computer: Password incorrect *1
Login to web vault with password: Password incorrect *1
Login to web vault with device: Notification comes, I approve, but still does not login. *2
Login to Android client with biometrics: Works
Login to Android client with password: Works

Both Opera and Chrome extensions are the latest version now: 2023.8.2

Here is the logs of failures:
Case *1: 
2023-09-05T06:21:11.429990371Z [2023-09-05 09:21:11.429][vaultwarden::api::identity][ERROR] Username or password is incorrect. Try again. IP: XXX.XXX.XXX.XXX. Username: XXXXX.

Case *2:
2023-09-05T06:19:33.884527856Z [2023-09-05 09:19:33.884][vaultwarden::api::core::accounts::_][WARN] Data guard `Json < AuthResponseRequest >` failed: Parse("{\"key\":\"xxxxxxxxxxxxxx==\",\"masterPasswordHash\":null,\"deviceIdentifier\":\"xxxxxxxxxxxxxxxxx\",\"requestApproved\":true}", Error("invalid type: null, expected a string", line: 1, column: 381)).

What have I recently changed?

- Installed the newest version
- Set push notification env vars (and it is working)
- Set password_iterations to the recommended value of 600000. (It was 100000 by default) (Reverting this doesn't help.)

The only effected user seems to be me (out of 15) and that's probably because; I first changed the iteration value in my web vault. Then I set it in admin interface. The other users' web vaults still show 100000 iterations (with a warning indicating it is very low and urging them to set it to a minimum of 600000).

Generated support string:

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.2
* Web-vault version: v2023.7.1
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": false,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*****************",
  "domain_origin": "*****://*****************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": 180,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": "***",
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "********",
  "invitations_allowed": false,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/data/bitwarden.log",
  "log_level": "warn",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": true,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": true,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": true,
  "signups_allowed": false,
  "signups_domains_whitelist": "***********",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "*******************",
  "smtp_from_name": "********",
  "smtp_host": "****************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "*******************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": 60,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": 100,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

**Comment by Turab** [comment]
@BlackDex , thank you for your time but I don't think this is a discussion. As you will see from the log I posted, this is most likely a bug:
Error("invalid type: null, expected a string", line: 1, column: 381)

**Reply by BlackDex** [reply]
See my comments below, it's probably not a bug. At least, not anymore for a few days already.

**Comment by BlackDex** [comment]
First, are you sure you can successfully login with a device via the web-vault using your master-password? Please also try a clean browser profile, or a browser which you hardly use and have no extensions installed on for example.

Second, the case 2 is solved if you update to the `testing` tagged images.

The setting of the iterations in the web-vault has nothing to do with the one from the Vaultwarden config. Those are different items. The web-vault is for the Bitwarden clients to know how much iterations they need to do. The Vaultwarden one is used for storing the master password hash.

Please also try to fully logout, or even uninstall the clients which do not work, and install and configure them again to see if that helps.

**Comment by Turab** [comment]
Now I tried to login to web vault with Edge (which I never use), still getting a password incorrect message. Fully logging out from the clients also didn't help.

Deployed testing image now, second case is resolved, login with device is now working, as you have said that seems to be a bug already fixed in the testing build. But case 1 prevails. I am using the same password to login to some clients but some other clients log it as password incorrect. Now I tried logging into web vault using a completely different device, but it still says password incorrect.

**Reply by BlackDex** [reply]
To be fair, it is almost totally impossible for one client to produce a bad password, and other clients to login successfully when using a master-password. The only thing i can think of is that something is either cached like the KDF iterations by the client which for some reason uses 100_000 iterations instead of 600_000, which in those cases would mean an invalid password.

If you have two browsers of which one works, and the other doesn't, i suggest to use the Developer Console (F12) and see what the difference is between the requests send. Or, if you are able to use something like wireshark of tcpflow on the Vaultwarden side/host to fetch the unencrypted traffic and check what the difference is between what the client sends, since the master-password hash should be exactly the same coming from any client.

**Comment by Turab** [comment]
To make it clear; I can't login to web vault from any browser on any device using the master password. I was able to login to Android client and to the Opera extension only. After deploying `testing` image per your suggestion, I am only now able to login to web vault but by only "device login" method. Master password still emits "wrong password" error.

So I can't test the difference between a successful login and a failing one, since I can never login to web vault with the master password. As for monitoring traffic with tcpflow, I couldn't manage to watch behind an nginx proxy. (Container is deployed o a Caprover instance)

**Reply by BlackDex** [reply]
I think something went wrong then with your conversion of the KDF iterations. That is the only thing i can think of.
I actually think that you then also arn't able to login with a mobile device either, but since you can just unlock currently you are still able to access your vault.

Please look in the database at your user-account and see what is set there for the `password_iterations`.
If that shows 600_000, and you reverted the server default to `PASSWORD_ITERATIONS=100000`, try to set the count for your account back to 100_000 and try to login.

I do recommend to create a backup of the database first.
Also, please check the logs of Vaultwarden if you see any errors around the time it broke for you.

I have no idea what `caprover` is, so no clue what you can or can't do there.
Again, the only thing i can think of is that there went something wrong with the password iterations, or you did some manual changes somewhere. Other thing i can suggest is to restore a backup of the database, and see if that works, and if that is the case, then see what the difference is for you specific user account.

**Comment by Turab** [comment]
I haven't checked what it showed on the DB because it was already 600000 on my account settings. (Was there a chance it could be different on the DB?) Both the server and my account's iterations were the same. (Other users are still 100000).

Since I already guessed it could be the problem, I mentioned it in my first post that it was the only thing I changed. Nothing edited manually. I think other users will have the same problem once they change their iteration settings to the recommended value.

Nevertheless, I couldn't risk loosing access to my vault and restoring a backup. I just created another account with another e-mail of mine and gave emergency access to this new account to let the new account set a new master password for my vault. I changed the master password with emergency access and now I can login to web vault with it.

Edit: I needed to log out of the Chrome extension before it let me login again. Otherwise it was still complaining about a wrong password. Plus, there is something new here which might be related to the issue. When I set a new password through emergency access, it forced me to setup a minimum 12 chars long password. My actual master password was 8 chars long. Could this be breaking the iteration? This makes it a bug I guess. (Of which the replicating steps would be: 1) Have a password less than 12 chars long. 2) Change account iteration to a new value. 3) Change iteration settings of the server to this new value. 4) Try logging to web vault.) This doesn't explain how some clients could login while some could not which is a big peculiarity.

**Reply by BlackDex** [reply]
Do you know what the previous version was you used before you updated?
Because the 600_000 iterations has been the default for a while already.

If i want to try and replicate this i need to know from which version you came.
Also, i think you are still mixing the iterations. The one in the account settings are client iterations, the one you can configure for Vaultwarden have nothing to do with those, only for storing the master-password hash, not for creating the master-password hash.

**Comment by Turab** [comment]
Here is my deploy history:
![image](https://github.com/dani-garcia/vaultwarden/assets/1780719/8cf63629-5f2c-4b8e-b460-30c763c12e6b)

**Reply by BlackDex** [reply]
Then, if you didn't set the `PASSWORD_ITERATIONS` your self, it should have used the default which has changed since v1.28.0.
So, did you changed that value manually? If yes, did you do that since the v1.29.2 version? And if so, did you changed the KDF iterations in your account afterwards?

Because what happens if  you change the `PASSWORD_ITERATIONS` to a different value then is in the database for a specific user, it will update the iterations during a login attempt with a password. It will check if the `password_iterations` in the database is different from the settings, if that is the case it will update this. Else it will leave it untouched.

The password/kdf iterations in the web-vault are only used by the clients, and are only stored in the database to provide this value so the clients know how much iterations they need to do. That value is stored in the `client_kdf_iter` column.

**Reply by Turab** [reply]
To be honest, I couldn't remember for sure if it was me who changed it on admin interface or it just took it by default after a version. But I remember setting it back to 100000 to see if it would have made a difference, which didn't.
